File: 프로그래머스/2/87946. 피로도/피로도.py
'''
1. 아이디어
    - 최소 피로도, 소모 피로도
    - 유저는 최소 피로도 이상을 갖고 > 던전 > 피로도 - 소모 피로도
    > 이어서 던전을 참가하기 위해서는 (기존 피로도 - 소모 피로도)가 다음 던전의 최소 피로도 이상이어야 한다
    - 최대 던전의 수를 return
    
    - for loop
'''
from itertools import permutations

def solution(k, dungeons):
    
    # 1. 가능한 모든 순서 정하기
    all_case = list(permutations(dungeons))
    
    # 2. 각 case들을 보면서, 얼마나 많이 갈 수 있는지 구하기
    total_count = set()
    
    for case in all_case:
        cnt = 0
        heart = k
        for dungeon in case:
            if heart >= dungeon[0]: # 현재, 최소피로도 이상일 경우 참여 가능
                heart -= dungeon[1]
                cnt += 1
            else:
                break
        total_count.add(cnt)
                
    # 3. 경우의 수 중 가장 큰 값 반환
    return max(total_count)
            
    

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = 0; 

    public int solution(int k, int[][] dungeons) {
        
        boolean[] visited = new boolean[dungeons.length];
        
        // 탐험 시작
        dfs(k, dungeons, visited, 0);
        
        return answer;
    }

    static void dfs(int k, int[][] dungeons, boolean[] visited, int cnt) {
        // 매 단계마다 현재까지 탐험한 던전 수의 최댓값을 갱신
        answer = Math.max(answer, cnt);

        for (int i = 0; i < dungeons.length; i++) {
            // 1. 방문하지 않았고 2. 현재 피로도가 최소 필요 피로도 이상인 경우
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true; // 방문 처리
                
                // 다음 던전 탐색
                dfs(k - dungeons[i][1], dungeons, visited, cnt + 1);
                
                visited[i] = false; // 복구 (백트래킹)
            }
        }
    }
}